File: js/distill.js
/*
 * distill.js — flow tensor → swatch parameters (carried from v1).
 */

function distillSwatch(frames, fps) {
  const T = frames.length;
  if (T < 16) return null;

  const gvx = new Float64Array(T);
  const gvy = new Float64Array(T);
  for (let t = 0; t < T; t++) {
    let sx = 0, sy = 0, n = 0;
    for (const p of frames[t]) { if (p.valid) { sx += p.vx; sy += p.vy; n++; } }
    gvx[t] = n ? sx / n : 0;
    gvy[t] = n ? sy / n : 0;
  }
  const mx = _mean(gvx), my = _mean(gvy);
  for (let t = 0; t < T; t++) { gvx[t] -= mx; gvy[t] -= my; }

  let cxx = 0, cxy = 0, cyy = 0;
  for (let t = 0; t < T; t++) { cxx += gvx[t] ** 2; cxy += gvx[t] * gvy[t]; cyy += gvy[t] ** 2; }
  const disc = Math.sqrt(Math.max(0, (cxx - cyy) ** 2 / 4 + cxy ** 2));
  const l1 = (cxx + cyy) / 2 + disc;
  let ax, ay;
  if (Math.abs(cxy) > 1e-9) { ax = l1 - cyy; ay = cxy; }
  else if (cxx >= cyy) { ax = 1; ay = 0; } else { ax = 0; ay = 1; }
  const alen = Math.hypot(ax, ay) || 1; ax /= alen; ay /= alen;
  let direction = Math.atan2(-ay, ax) * 180 / Math.PI;
  if (direction < 0) direction += 180;
  if (direction >= 180) direction -= 180;

  const proj = new Float64Array(T);
  for (let t = 0; t < T; t++) proj[t] = gvx[t] * ax + gvy[t] * ay;

  const { spectrum, peakBin } = _dft(proj);
  const binHz = fps / T;
  let freqHz = peakBin * binHz;
  if (peakBin > 1 && peakBin < spectrum.length - 1) {
    const a = Math.log(spectrum[peakBin - 1] + 1e-12);
    const b = Math.log(spectrum[peakBin] + 1e-12);
    const c = Math.log(spectrum[peakBin + 1] + 1e-12);
    const d = 0.5 * (a - c) / (a - 2 * b + c || 1e-12);
    if (Math.abs(d) < 1) freqHz = (peakBin + d) * binHz;
  }
  freqHz = Math.max(0.05, Math.min(8, freqHz));

  let rms = 0;
  for (let t = 0; t < T; t++) rms += gvx[t] ** 2 + gvy[t] ** 2;
  rms = Math.sqrt(rms / T);
  const amplitude = _clamp(rms / 1.5);

  const flatness = _spectralFlatness(spectrum);
  const disorder = _spatialDisorder(frames);
  const turbulence = _clamp(0.5 * flatness + 0.5 * disorder);

  const damping = _autocorrDecay(proj, fps, freqHz);
  const phaseSpread = _phaseSpread(frames, ax, ay, peakBin);

  return {
    frequency: _r3(freqHz), amplitude: _r3(amplitude), direction: Math.round(direction),
    turbulence: _r3(turbulence), damping: _r3(damping), phaseSpread: _r3(phaseSpread),
  };
}

function _mean(a) { let s = 0; for (const v of a) s += v; return s / a.length; }
function _clamp(v) { return Math.max(0, Math.min(1, v)); }
function _r3(v) { return Math.round(v * 1000) / 1000; }

function _dft(signal) {
  const T = signal.length, half = T >> 1;
  const spectrum = new Float64Array(half + 1);
  let peakBin = 1, peakMag = -1;
  for (let k = 1; k <= half; k++) {
    let re = 0, im = 0;
    for (let t = 0; t < T; t++) { const ph = -2 * Math.PI * k * t / T; re += signal[t] * Math.cos(ph); im += signal[t] * Math.sin(ph); }
    spectrum[k] = Math.hypot(re, im);
    if (spectrum[k] > peakMag) { peakMag = spectrum[k]; peakBin = k; }
  }
  return { spectrum, peakBin };
}

function _spectralFlatness(spectrum) {
  let logSum = 0, sum = 0, n = 0;
  for (let k = 1; k < spectrum.length; k++) { const p = spectrum[k] ** 2 + 1e-12; logSum += Math.log(p); sum += p; n++; }
  return n ? _clamp(Math.exp(logSum / n) / (sum / n)) : 0;
}

function _spatialDisorder(frames) {
  let acc = 0, cnt = 0;
  for (const frame of frames) {
    let cx = 0, cy = 0, wsum = 0;
    for (const p of frame) { if (!p.valid) continue; const m = Math.hypot(p.vx, p.vy); if (m < 0.05) continue; cx += p.vx / m * m; cy += p.vy / m * m; wsum += m; }
    if (wsum < 0.3) continue;
    acc += 1 - Math.hypot(cx, cy) / wsum; cnt++;
  }
  return cnt ? acc / cnt : 0;
}

function _autocorrDecay(signal, fps, freqHz) {
  const T = signal.length;
  const lag = Math.max(2, Math.min(T - 4, Math.round(fps / freqHz)));
  let num = 0, den = 0;
  for (let t = 0; t < T - lag; t++) num += signal[t] * signal[t + lag];
  for (let t = 0; t < T; t++) den += signal[t] * signal[t];
  return den < 1e-9 ? 0.5 : _clamp(1 - Math.abs(num / den));
}

function _phaseSpread(frames, ax, ay, k) {
  const T = frames.length, N = frames[0].length;
  let cx = 0, cy = 0, n = 0;
  for (let i = 0; i < N; i++) {
    let re = 0, im = 0, energy = 0;
    for (let t = 0; t < T; t++) { const p = frames[t][i]; const s = p.valid ? p.vx * ax + p.vy * ay : 0; const ph = -2 * Math.PI * k * t / T; re += s * Math.cos(ph); im += s * Math.sin(ph); energy += s * s; }
    if (energy / T < 0.005) continue;
    const mag = Math.hypot(re, im) || 1; cx += re / mag; cy += im / mag; n++;
  }
  return n < 4 ? 0.3 : _clamp(1 - Math.hypot(cx, cy) / n);
}

window.distillSwatch = distillSwatch;

File: service/distill.py
"""Distill dense flow into the 6 motion params + a GRIDxGRID trajectory grid."""
import math
import numpy as np
from config import GRID


# ---------------------------------------------------------------- distill
def _clamp01(v):
    return float(max(0.0, min(1.0, v)))


def distill(flows: np.ndarray, fps: float, noise_floor: float = 0.55,
            mask: np.ndarray = None) -> dict:
    """Dense-flow port of the browser's distillSwatch (see js/distill.js).

    flows: [T, 2, H, W] px/frame.
    noise_floor: px/frame below which residual flow is treated as RAFT noise.
        Default 0.55 is calibrated on benchmark.py's static clip for whole-frame
        analysis. segment_regions() passes a lower value for atmospheric motion
        (rain, smoke, mist) whose real residual flow is ~0.08 px/frame — well
        below the default floor — so it would otherwise register amplitude 0.
    mask: optional boolean [H, W]. When given, ALL whole-frame statistics
        (drift mean, moving fraction, magnitude persistence, spatial disorder)
        are computed over the masked pixels only. segment_regions() passes a
        region's true (non-rectangular) pixel mask here — WITHOUT this, a
        sparse/diagonal region's zeroed bbox pixels read as static texture,
        inflating persistence and crushing the region's amplitude to ~0 purely
        as a function of its shape. Default None = whole frame (byte-identical
        to the pre-mask behaviour for every existing caller).
    """
    T, _, H, W = flows.shape
    if T < 12:
        return None

    # pixel selector: mask flattens [H,W]→[Npix] over real pixels only; when no
    # mask is given, all pixels participate (identical to the original code).
    if mask is not None:
        sel = mask.reshape(-1).astype(bool)
        if int(sel.sum()) < 4:
            return None
    else:
        sel = None

    def _region(px):
        """[..., H, W] → [..., Npix] over the mask (or flattened whole frame)."""
        flat = px.reshape(px.shape[:-2] + (H * W,))
        return flat[..., sel] if sel is not None else flat

    # global per-frame mean flow; keep the DC component as DRIFT (constant
    # travel: falling leaves, flowing water) before removing it so the
    # oscillation analysis below sees only the periodic part
    g_raw = _region(flows).mean(axis=2)      # [T, 2] mean over region pixels
    drift_px = g_raw.mean(axis=0)            # mean flow = steady drift
    # normalize: 0.5% of frame width per frame is a fast drift → ±1
    drift_x = float(np.clip(drift_px[0] / (0.005 * W), -1, 1))
    drift_y = float(np.clip(drift_px[1] / (0.005 * W), -1, 1))
    g = g_raw - g_raw.mean(axis=0, keepdims=True)
    gvx, gvy = g[:, 0], g[:, 1]

    # dominant axis via covariance eigenvector
    cov = np.cov(np.stack([gvx, gvy]))
    evals, evecs = np.linalg.eigh(cov)
    ax, ay = evecs[:, int(np.argmax(evals))]
    direction = math.degrees(math.atan2(-ay, ax)) % 180.0

    proj = gvx * ax + gvy * ay

    # dominant frequency via rFFT with parabolic refinement
    spec = np.abs(np.fft.rfft(proj))
    spec[0] = 0.0
    k = int(np.argmax(spec))
    freq = k * fps / T
    if 1 <= k < len(spec) - 1:
        la, lb, lc = (math.log(spec[k - 1] + 1e-12),
                      math.log(spec[k] + 1e-12),
                      math.log(spec[k + 1] + 1e-12))
        denom = la - 2 * lb + lc
        if abs(denom) > 1e-12:
            d = 0.5 * (la - lc) / denom
            if abs(d) < 1:
                freq = (k + d) * fps / T
    freq = float(min(8.0, max(0.05, freq)))

    # amplitude of the OSCILLATING part = how strongly the MOVING pixels move.
    # Whole-frame averaging dilutes sparse movers (a flag on sky, scattered
    # leaves); global-mean flow additionally cancels incoherent motion. So:
    # mean magnitude over pixels above RAFT's static-scene noise floor
    # (~0.2-0.4 px/frame, calibrated on benchmark.py's static clip), scaled
    # by a soft coverage factor so pure-noise frames stay near zero.
    NOISE_FLOOR = noise_floor
    resid = flows - drift_px.reshape(1, 2, 1, 1)
    mag_map = np.sqrt(resid[:, 0] ** 2 + resid[:, 1] ** 2)   # [T, H, W]
    mag_reg = _region(mag_map)                               # [T, Npix] (region only)
    moving = mag_reg > NOISE_FLOOR
    frac = float(moving.mean())
    # RAFT hallucinates 0.3-0.9 px/frame even on static scenes, but that
    # noise is ANCHORED to texture edges: the spatial magnitude pattern is
    # ~identical frame to frame (Pearson corr ≈ 1.0), while real motion
    # travels through space (measured ≤ 0.7 on every moving benchmark clip
    # and both real videos). Gate amplitude by 1 - persistence. Computed over
    # region pixels only — otherwise the zeroed non-region pixels of a sparse
    # crop read as static texture and falsely inflate persistence.
    a = mag_reg[:-1]
    b = mag_reg[1:]
    am = a - a.mean(axis=1, keepdims=True)
    bm = b - b.mean(axis=1, keepdims=True)
    corr = (am * bm).sum(axis=1) / (np.sqrt((am**2).sum(axis=1) * (bm**2).sum(axis=1)) + 1e-9)
    persistence = float(corr.mean())
    static_gate = _clamp01((0.92 - persistence) / 0.12)   # 1 below 0.80, 0 above 0.92
    if frac > 0.003:
        strength = float((mag_reg[moving] - NOISE_FLOOR).mean())
        coverage = min(1.0, frac / 0.03)          # full weight at ≥3% of pixels
        amplitude = _clamp01(strength / (0.006 * W) * coverage * static_gate)
    else:
        amplitude = 0.0

    # turbulence: spectral flatness x spatial disorder
    p = spec[1:] ** 2 + 1e-12
    flatness = float(np.exp(np.log(p).mean()) / p.mean())
    mag = np.sqrt(flows[:, 0] ** 2 + flows[:, 1] ** 2) + 1e-9   # [T, H, W]
    ux, uy = flows[:, 0] / mag, flows[:, 1] / mag
    # region-only weighted resultant (spatial disorder of flow directions)
    mag_r = _region(mag)                    # [T, Npix]
    ux_r, uy_r = _region(ux), _region(uy)
    npix = mag_r.shape[1]
    wsum = mag_r.sum(axis=1)
    rx = (ux_r * mag_r).sum(axis=1) / wsum
    ry = (uy_r * mag_r).sum(axis=1) / wsum
    moving = wsum / npix > 0.02             # frames with real motion
    disorder = float((1 - np.sqrt(rx**2 + ry**2))[moving].mean()) if moving.any() else 0.0
    # 1.4x stretch: RAFT (esp. raft_large) produces spatially smooth flow, which
    # compresses the disorder statistic; calibrated on benchmark.py clips
    turbulence = _clamp01((0.5 * flatness + 0.5 * disorder) * 1.4)

    # damping: autocorrelation at one dominant-period lag
    lag = max(2, min(T - 4, round(fps / freq)))
    den = float((proj**2).sum())
    r = abs(float((proj[:-lag] * proj[lag:]).sum()) / den) if den > 1e-9 else 0.5
    damping = _clamp01(1 - r)

    # phase spread: energy-weighted circular variance of PER-PIXEL phase at
    # the dominant bin. Cell-averaging blends opposite-phase pixels within a
    # cell (cancellation) and RAFT's smoothness prior bleeds flow into the
    # background, so we go per-pixel: single-bin DFT via dot product with
    # exp(-2πikt/T), weight each pixel by its spectral energy at that bin.
    pix_proj = flows[:, 0] * ax + flows[:, 1] * ay           # [T, H, W]
    kk = max(k, 1)
    basis = np.exp(-2j * np.pi * kk * np.arange(T) / T).astype(np.complex64)
    ft = np.tensordot(basis, pix_proj.astype(np.float32), axes=(0, 0))  # [H, W] complex
    ft_flat = ft.reshape(-1)
    ft_reg = ft_flat[sel] if sel is not None else ft_flat    # region pixels only
    w = (np.abs(ft_reg) ** 2).astype(np.float64)
    if float(w.sum()) > 1e-9:
        unit = ft_reg / (np.abs(ft_reg) + 1e-12)
        resultant = np.abs((unit * w).sum()) / w.sum()
        phase_spread = _clamp01(1 - float(resultant))
    else:
        phase_spread = 0.3

    # a near-static clip yields noise-driven turbulence/phase estimates, and
    # turbulence displacement is not amplitude-gated in the renderer — gate
    # them here so "no motion in" ⇒ "no motion out"
    gate = min(1.0, amplitude / 0.10)
    turbulence *= gate
    phase_spread *= gate

    return {
        "frequency": round(freq, 3),
        "amplitude": round(amplitude, 3),
        "direction": round(direction),
        "turbulence": round(turbulence, 3),
        "damping": round(damping, 3),
        "phaseSpread": round(phase_spread, 3),
        # steady directional travel (−1..1 of a fast drift), e.g. falling
        # leaves ≈ (0, +0.x); oscillation-only clips ≈ (0, 0)
        "driftX": round(drift_x, 3),
        "driftY": round(drift_y, 3),
    }


def grid_trajectories(flows: np.ndarray) -> list:
    """Integrate mean cell flow into GRID x GRID point tracks (normalized coords)."""
    T, _, H, W = flows.shape
    ch, cw = H // GRID, W // GRID
    cells = flows[:, :, : ch * GRID, : cw * GRID].reshape(T, 2, GRID, ch, GRID, cw).mean(axis=(3, 5))
    tracks = []
    for gy in range(GRID):
        for gx in range(GRID):
            x = (gx + 0.5) / GRID
            y = (gy + 0.5) / GRID
            pts, px, py = [[round(x, 4), round(y, 4)]], x, y
            for t in range(T):
                px += float(cells[t, 0, gy, gx]) / W
                py += float(cells[t, 1, gy, gx]) / H
                pts.append([round(px, 4), round(py, 4)])
            tracks.append(pts)
    return tracks
